core/data/pack/reader.go

```diff
@@ -101,34 +88,50 @@ func (r *reader) unmarshal(ctx context.Context) error {
 		if err = r.pb.Unmarshal(desc); err != nil {
 			return err
 		}
-		if _, ok := r.types.addNameAndDesc(name, desc); !ok {
-			return ErrUnknownType{name}
-		}
+		r.types.add(name, desc)
 		return nil
+	}
+
+	// Read first two fields of object instance. If missing, they are implicitly set to 0.
+	// NB: Protobuf library returns the signed zig-zag-encoded integers as uint64!
+	parent, err := r.pb.DecodeZigzag64()
+	if err != nil && err != io.ErrUnexpectedEOF {
+		return err
+	}
+	tyIdx, err := r.pb.DecodeZigzag64()
+	if err != nil && err != io.ErrUnexpectedEOF {
+		return err
+	}
+	hasParent := int64(parent) < 0
+	hasChildren := int64(tyIdx) < 0
 
-	default:
-		tyIdx, isGroup := tyIdxAndGroupFromTag(tag)
+	if tyIdx == 0 { // Null-terminator
+		if hasParent {
+			if err := r.events.EndGroup(ctx, r.id+parent); err != nil {
+				return err
+			}
+		}
+	} else { // New object instance
+		if int64(tyIdx) < 0 {
+			tyIdx = -tyIdx // Absolute value.
```

Comment:
TIL: I did not know you could negate an unsigned value.

Reply:
Another fun fact: Did you know that: -MIN_INT == MIN_INT